doc/manual/src/architecture/architecture.md

```diff
@@ -0,0 +1,78 @@
+# Architecture
+
+This chapter describes how Nix works.
+It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.
+
+## Overview
+
+Nix consists of [hierarchical layers][layer-architecture].
+
+```
++-----------------------------------------------------------------+
+| Nix                                                             |
+|                  [ commmand line interface ]------,             |
+|                               |                   |             |
+|                           evaluates               |             |
+|                               |                manages          |
+|                               V                   |             |
+|                  [ configuration language  ]      |             |
+|                               |                   |             |
+| +-----------------------------|-------------------V-----------+ |
+| | store                  evaluates to                         | |
+| |                             |                               | |
+| |             referenced by   V       builds                  | |
+| |  [ build input ] ---> [ build plan ] ---> [ build result ]  | |
```

Comment:
```suggestion
| |  [ build input ] ---> [ build plan ] ---> [ outputs ]       | |
```

Isn't "output" the technical term for this, or should "result" encompass more than just the output? If so, what?
Example: `nix-store -q --outputs`

We should have a small glossary to which we then limit our use of terminology.

I wish markdown has `set -u` :wink:  (`nounset`)

Reply:
> Isn't "output" the technical term for this

It's the Nix term for this, as is *derivation* for what we call *build task* here. We deliberately use generic terms in the overview, as none of that is specific to Nix.

> We should have a small glossary to which we then limit our use of terminology.

We have that as the [Rosetta Stone](https://github.com/NixOS/nix/pull/6420/files#diff-485e9f03f514869c78ebfa54354bd03b1fd0ba9ceb817578f31b573be303b5f4R181). The idea is to introduce Nix specific terminology in the subsequent sections, where we also describe how Nix implements the concepts in principle. We could move the table to the overview to accommodate newcomers and experienced Nix users, see https://github.com/NixOS/nix/pull/6420#discussion_r919878416.

Reply:
> as none of that is specific to Nix.

Then `Nix` should be `package manager` and `store` should be `data and process layer`.

> Rosetta Stone

This doesn't scale to encompass terminology that is specific to Nix, so we should still have a glossary.

Would it make sense to have a translation of the entire diagram? Seems like a good way to learn the new terminology.

Reply:
> Then `Nix` should be `package manager` and `store` should be `data and process layer`.

Sounds reasonable, please add a suggestion, considering the following:
- `data and process layer` reads a bit clunky, although it's the correct term
  - should we call it "store" generically instead, in the table?
- how do we still clarify this diagram refers to how Nix does things

> This doesn't scale to encompass terminology that is specific to Nix, so we should still have a glossary.

I hope this would be covered by additional sections, see [the colored outline comment above](https://github.com/NixOS/nix/pull/6420#issuecomment-1161690195). Is there anything that would still be missing?
Not against a glossary, it just needs enough substance. Maybe abbreviations such as drv or FSO, which we currently don't use and should not use in the official writing, but people may find helpful to decipher. On the other hand those may be NixOS Wiki material.

> Would it make sense to have a translation of the entire diagram? Seems like a good way to learn the new terminology.

Would take too much head space for my taste. Or maybe just link to the respective pages behind the table, not mentioning it in the table of contents? Sound like detail work though, not relevant for this PR.

Reply:
Humans are really good at memorizing things in places, so I would guess that familiarizing oneself with the general diagram and then learning the technical terms in the same "place" would be highly effective. (I'm not a cognitive psychologist though)

You'd have to put them close together for this correspondence to be noticed; not one at the start and one at the end.

> Is there anything that would still be missing?

 - Terminology that is more specific, like "store path", "store hash", "store directory".
 - Terminology that builds on top of the core architecture, such as "input", which has a very specific meaning in Nixpkgs

One could argue that the latter is out of scope for this document, but in practice, all terminology throughout the ecosystem shares the same namespace. We should recognize that and have a single place to look all of them up, and disambiguate them where necessary. (e.g. "input", which is already conflicting with a term used in this chapter)

 > how do we still clarify this diagram refers to how Nix does things?
 
 The context would imply this, but it'd be helpful to have a caption such as "An overview of the Nix architecture using generic terminology", which then contrasts well with "An overview of the Nix architecture using its specific terms".

Reply:
> Terminology that is more specific, like "store path", "store hash", "store directory".

We have that in the table of contents, and it is entirely Nix-specific.

> One could argue that the latter is out of scope for this document, but in practice, all terminology throughout the ecosystem shares the same namespace.

Agreed on potential confusion. Suggestions to do anything about that except always referring to "build input" in generic terms, because that is different from just "input"?

> You'd have to put them close together for this correspondence to be noticed; not one at the start and one at the end.

I agree with the learning conditions premise, but cannot picture a specific change. Feel free to make a pull request [on the fork](https://github.com/nix-community/) against this branch.